server/rest/src/main/java/org/infinispan/rest/resources/CacheResource.java

```diff
@@ -123,95 +119,80 @@ private RestResponse deleteCacheValue(RestRequest request) throws RestResponseEx
+      });
    }
 
-   private RestResponse putValueToCache(RestRequest request) {
-      try {
-         String cacheName = request.variables().get("cacheName");
+   private CompletionStage<RestResponse> putValueToCache(RestRequest request) {
 
-         MediaType contentType = request.contentType();
-         MediaType keyContentType = request.keyContentType();
+      String cacheName = request.variables().get("cacheName");
 
-         AdvancedCache<Object, Object> cache = restCacheManager.getCache(cacheName, keyContentType, contentType);
-         Object key = request.variables().get("cacheKey");
-         if (key == null) throw new NoKeyException();
-         NettyRestResponse.Builder responseBuilder = new NettyRestResponse.Builder();
+      MediaType contentType = request.contentType();
+      MediaType keyContentType = request.keyContentType();
 
-         if (request.method() == POST && cache.containsKey(key)) {
-            return responseBuilder.status(HttpResponseStatus.CONFLICT.code()).entity("An entry already exists").build();
-         } else {
-            ContentSource contents = request.contents();
-            if (contents == null) throw new NoDataFoundException();
-            byte[] data = request.contents().rawContent();
-            CacheEntry<Object, Object> entry = restCacheManager.getInternalEntry(cacheName, key, true, keyContentType, contentType);
-            if (entry instanceof InternalCacheEntry) {
-               InternalCacheEntry ice = (InternalCacheEntry) entry;
-               String etagNoneMatch = request.getEtagIfNoneMatchHeader();
-               if (etagNoneMatch != null) {
-                  String etag = calcETAG(ice.getValue());
-                  if (etagNoneMatch.equals(etag)) {
-                     //client's and our ETAG match. Nothing to do, an entry is cached on the client side...
-                     responseBuilder.status(HttpResponseStatus.NOT_MODIFIED.code());
-                     return responseBuilder.build();
-                  }
+      AdvancedCache<Object, Object> cache = restCacheManager.getCache(cacheName, keyContentType, contentType);
+      Object key = request.variables().get("cacheKey");
+      if (key == null) throw new NoKeyException();
+      NettyRestResponse.Builder responseBuilder = new NettyRestResponse.Builder();
+
+      if (request.method() == POST && cache.containsKey(key)) {
+         return CompletableFuture.completedFuture(responseBuilder.status(HttpResponseStatus.CONFLICT.code()).entity("An entry already exists").build());
+      }
+      ContentSource contents = request.contents();
+      if (contents == null) throw new NoDataFoundException();
+      Long ttl = request.getTimeToLiveSecondsHeader();
+      Long idle = request.getMaxIdleTimeSecondsHeader();
+
+      byte[] data = request.contents().rawContent();
+
+      return restCacheManager.getInternalEntry(cacheName, key, true, keyContentType, contentType).thenComposeAsync(entry -> {
```

Comment:
Looks like the code in here is non blocking, so this should be `thenCompose` instead.

Reply:
Also just to note if we use *Async method we should use an explicit Executor.

Reply:
leftover, thanks for spotting it, it's already using ```thenCompose``` elsewhere and not need to use async methods